qa/tasks/caasp.py

```diff
@@ -0,0 +1,100 @@
+'''
+Task that deploys a CAASP cluster on all the nodes
+Linter:
+    flake8 --max-line-length=100
+'''
+import logging
+from teuthology.misc import (
+    all_roles_of_type,
+    append_lines_to_file,
+    get_file
+    )
+from teuthology.task import Task
+from util import (
+    get_remote_for_role,
+    )
+log = logging.getLogger(__name__)
+
+
+class Caasp(Task):
+    """
+    Deploy a Caasp cluster on all remotes using Skuba.
+    Nodes are declared in suites/caasp folder
+    """
+
+    def __init__(self, ctx, config):
+        super(Caasp, self).__init__(ctx, config)
+        log.debug("beginning of constructor method")
+        self.ctx['roles'] = self.ctx.config['roles']
+        self.log = log
+        self.remotes = self.cluster.remotes
+        self.mgmt_remote = get_remote_for_role(self.ctx, "skuba_mgmt_host.0")
+        self.ssh_priv = 'caasp_key.rsa'
+        self.ssh_pub = 'caasp_key.rsa.pub'
+        self.set_agent = "eval $(ssh-agent) && ssh-add ~/{} && ".format(
+            self.ssh_priv)
+
+    def __ssh_setup(self):
+        """ Generate keys on management node. Copy pub to all of them. """
+        log.debug("Executing SSH setup")
+        self.__ssh_gen_key()
+        self.__ssh_copy_pub_to_caasp()
+
+    def __ssh_gen_key(self):
+        self.mgmt_remote.run(args=[
+            'ssh-keygen',
+            '-t',
+            'rsa',
+            '-b',
+            '2048',
+            '-N',
+            '',
+            '-f',
+            '{}'.format(self.ssh_priv),
+        ])
+
+    def __ssh_copy_pub_to_caasp(self):
+        log.debug("Copying public key to remotes")
+        data = get_file(self.mgmt_remote, self.ssh_pub)
+        for i, _ in enumerate(all_roles_of_type(self.ctx.cluster, 'caasp_master')):
+            r = get_remote_for_role(self.ctx, 'caasp_master.' + str(i))
+            append_lines_to_file(r, '.ssh/authorized_keys', data)
+        for i, _ in enumerate(all_roles_of_type(self.ctx.cluster, 'caasp_worker')):
+            r = get_remote_for_role(self.ctx, 'caasp_worker.' + str(i))
+            append_lines_to_file(r, '.ssh/authorized_keys', data)
+
+    def __create_cluster(self):
+        log.debug('Creating cluster')
+        master_remote = get_remote_for_role(self.ctx, "caasp_master.0")
+        commands = [
+            "ssh-add -L",
+            "skuba cluster init --control-plane {} cluster".format(master_remote.hostname),
+            "cd cluster && skuba node bootstrap --user ubuntu --sudo --target {} my-master".format(
+                master_remote.hostname),
+        ]
+        for command in commands:
+            self.mgmt_remote.sh("%s %s" % (self.set_agent, command))
```

Comment:
To avoid multiple ssh connection, improve code readability and avoid code duplication it is better to use this code as a single block, like:
```
self.mgmt_remote.sh(self.set_agent + """
    ssh-add -L
    skuba cluster init --control-plane {master} cluster
    cd cluster
    skuba node bootstrap --user {user} --sudo --target {master} my-master
    """.format(master=master_remote.hostname, user='ubuntu')
```

Reply:
Is it really a problem to do multiple SSH connections ?
AFAIK I prefer to have a loop so that we can easily add steps and also more easily see which step is failing

Reply:
In general less ssh connection you use is better, the code will go faster when you run it as snippets. And it is easier to read and maintain, when you read a solid piece of shell code instead of strange array of array of array combined with escaped code of escape symbols. In general loop are good but for there purpose, but using it for splitting up shell script it is not the best choice to improve error visibility, for example, when you do:
```
for i in ["some_command", "some_command"]:
    remote.run(args=i)
```
It is difficult to say which of the same command failed, while analyzing logs.
So it is better to have a separate call, so you one see the stack trace so you can find a line in python script or, use a code snippet, so you can see the line number within the script by bash or shell.